### PR TITLE
Fix avatar attachment tearing

### PR DESCRIPTION
**Pull request by samcake**
This PR fixes the tearing happening on the attachments of an avatar.

In the model, the MeshState(s) contains the current values for the cluster matrices and the cluster buffer used for rendering.
The cluster matrices values are updated correctly in Game loop (good)
But these values were copied into the ClusterBuffer also in Game loop (bad)
The cluster buffer is used in Render loop, assigned to the batch for rendering (good)
And worse, the cluster matrices were read from Game Loop with potential threading issues (very bad)

The bug was experienced in heavy loaded domain because in that case, the values in cluster matrices and the one in cluster buffer may likely be different as game loop and render loop are running at different rates.

To address this bug, we are making sure that:
- cluster matrices is still updated in Game loop
- the new values for the cluster matrices are communicated through a transaction to the Scene Render Items so Cluster Buffer(s) only exist in the Render loop side (not anymore in the MeshState)
- we do not need to access the model.state.ClusterMatrices from the render loop anymore

This guaranty thread safety for the cluster matrices between Game loop and Render loop.
All objects relying on the same Rig / cluster matrices values from one Game frame will be displayed using these same values in the same Render frame

## TEST PLAN
THe repro case is explained in this bug:
https://highfidelity.fogbugz.com/f/cases/7628/Parented-Entities-Moves-Are-Delayed-From-Parent

Repro:
0. in dev-welcome
1. Use the Robimo Frost avatar
2. Wear this attachment on your hip joint: http://west.nexusheavy.com/athauxan/Tinybot/WonkaSuit/wonkasuit_05.fbx
Run this script from the script console:
`(function () {
	'use strict';
	var main = function () {
		var attachment = {
		modelURL: "http://west.nexusheavy.com/athauxan/Tinybot/WonkaSuit/wonkasuit_05.fbx",
		jointName: "Hips",
		translation: {"x": 0, "y": 0.0, "z": 0},
		rotation: {"x": 0, "y": 0, "z": 0, "w": 1},
		scale: 1.0,
		isSoft: true
		};	
		MyAvatar.attach(attachment.modelURL,
					 attachment.jointName,
					 attachment.translation,
					 attachment.rotation,
					 attachment.scale,
					 attachment.isSoft);
		Script.stop();
	};
	main();
})();
`
3. Walk/jump around
Observed: the avatar body clips through the coat due to a slight lag in the attachment's movement relative to the avatar

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4669/HighFidelity-Beta-PR11742-390f5b53caeaab609bddebefc0eb8a54de6ea9d5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4669/HighFidelity-Beta-PR11742-390f5b53caeaab609bddebefc0eb8a54de6ea9d5.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4681/HighFidelity-Beta-PR11742-053052a95db484cb48a55a4584d6c939bbb8b2e2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4685/HighFidelity-Beta-PR11742-7da69633f36fa244c9a0daabae6a263ad7bb1034.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4685/HighFidelity-Beta-PR11742-7da69633f36fa244c9a0daabae6a263ad7bb1034.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4718/HighFidelity-Beta-PR11742-f665bf384b30ca82feacf9978c0126ad592266b4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4718/HighFidelity-Beta-PR11742-f665bf384b30ca82feacf9978c0126ad592266b4.exe

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

[Was unable to access Dev-welcome on current master build 7360, used Welcome in its place]
Use the Robimo Frost avatar **Check**
Wear this attachment on your hip joint: http://west.nexusheavy.com/athauxan/Tinybot/WonkaSuit/wonkasuit_05.fbx **Check, but not without issues.... Please see below image**
image

Walk/jump around **Check, but attachment not working correctly so not possible to judge**
Observed: the avatar body clips through the coat due to a slight lag in the attachment's movement relative to the avatar **Unable to confirm due to above**

using this PR I was unable to even add a  new attachment, the button shown in image would not do anything, tried restarting interface but same results.
image


**Fail**

**Comment by samcake** [comment]
right, i think this has to do with a new bug in master not related to this PR
you could use a wearable script (like a hat) and change the url in the interface after wards

**Comment by samcake** [comment]
we have found a bug with cauterized case (when viewed from first person)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4726/HighFidelity-Beta-PR11742-5ca3113709904fd25e4034cf8bd07a820bef40f9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4726/HighFidelity-Beta-PR11742-5ca3113709904fd25e4034cf8bd07a820bef40f9.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4729/HighFidelity-Beta-PR11742-6481e0f713169e51205855ff64688b32b86b019e.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4729/HighFidelity-Beta-PR11742-6481e0f713169e51205855ff64688b32b86b019e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4738/HighFidelity-Beta-PR11742-0718763ed1d8ff076e13ca8de8460236bd365d80.dmg

**Comment by marko8904** [comment]
build this please

**Comment by Nex-Pro** [comment]
@marko8904 Should I wait for the fresh build to test this ?

**Comment by Nex-Pro** [comment]
NM I'll wait.

**Comment by marko8904** [comment]
@Nex-Pro I'll have to try again. build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4746/HighFidelity-Beta-PR11742-0718763ed1d8ff076e13ca8de8460236bd365d80.exe

**Comment by Nex-Pro** [comment]
### Pretty sure this is a pass.

There is some minor poke-through near the elbows  ( See front gif) this is due to a minor rigging issue IMO.

@samcake I'll assume that the hoped-for outcome of this PR is that there is no longer any poke-through of the avatar.

**Currently avatar attachments are not working as indicated by @YuppyPlays  up the page. Not due to this pr.**

***I hacked a script to get the coat on the avatar in order to get this tested.***

### See attached gifs for outcome.

hifi-snap-by--on-2017-11-08_13-58-14
hifi-snap-by--on-2017-11-08_13-57-02
hifi-snap-by--on-2017-11-08_13-57-27
hifi-snap-by--on-2017-11-08_13-57-47

**Comment by Nex-Pro** [comment]
@samcake 

filed https://highfidelity.fogbugz.com/f/cases/9080/Avatar-attachments-dialog-not-working

**Comment by samcake** [comment]
Build this please

**Comment by samcake** [comment]
@Nex-Pro Thanks and thanks

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4757/HighFidelity-Beta-PR11742-0718763ed1d8ff076e13ca8de8460236bd365d80.dmg

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4757/HighFidelity-Beta-PR11742-0718763ed1d8ff076e13ca8de8460236bd365d80.exe

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp [hunk 1]
```diff
@@ -728,19 +728,19 @@ void Avatar::simulateAttachments(float deltaTime) {
         glm::quat jointRotation;
         if (attachment.isSoft) {
             // soft attachments do not have transform offsets
-            model->setTranslation(getPosition());
-            model->setRotation(getOrientation() * Quaternions::Y_180);
+            model->setTransformNoUpdateRenderItems(Transform(getOrientation() * Quaternions::Y_180, glm::vec3(1.0), getPosition()));
             model->simulate(deltaTime);
+            model->updateRenderItems();
```
Comment (AndrewMeadows) [inline libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp]: I had forgotten about `Model::setTransformNoUpdateRenderItems()`.  Maybe we should remove the "`temporary HACK`" comment about that method in the header file?

File: libraries/render-utils/src/CauterizedMeshPartPayload.cpp [hunk 2]
```diff
@@ -20,11 +20,32 @@ using namespace render;
...
+    // Once computed the cluster matrices, update the buffer(s)
+    if (clusterMatrices.size() > 1) {
+        if (!_clusterBuffer) {
+            _clusterBuffer = std::make_shared<gpu::Buffer>(clusterMatrices.size() * sizeof(glm::mat4),
+            (const gpu::Byte*) clusterMatrices.data());
+        } else {
+            _clusterBuffer->setSubData(0, clusterMatrices.size() * sizeof(glm::mat4),
+            (const gpu::Byte*) clusterMatrices.data());
+        }
+    }
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/CauterizedMeshPartPayload.cpp]: This upper part appears to be a duplicate implementation of `ModelMeshPartPayload::updateClusterBuffer()`.  Maybe just call the base class implementation here instead?

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp (hunk 1 above)
Comment (samcake) [reply libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp]: i m going to keep getting rid of the setOrientation and setTranslation on the model befoere changing it s status

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4763/HighFidelity-Beta-PR11742-75ce61c3a6950e4bc252e454616b8278b8ff90e1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4766/HighFidelity-Beta-PR11742-af1180255ead9d31ea88e9c23ee313f5fe339429.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4766/HighFidelity-Beta-PR11742-af1180255ead9d31ea88e9c23ee313f5fe339429.exe

**Pull request merged**